libraries/AP_WheelEncoder/AP_WheelEncoder.cpp

```diff
@@ -0,0 +1,239 @@
+        uint8_t type = _type[num_instances];
+        uint8_t instance = num_instances;
+
+        if (type == WheelEncoder_TYPE_QUADRATURE) {
+            state[instance].instance = instance;
+            drivers[instance] = new AP_WheelEncoder_Quadrature(*this, instance, state[instance]);
+        }
+#endif
+
+        if (drivers[i] != nullptr) {
+            // we loaded a driver for this instance, so it must be
+            // present (although it may not be healthy)
+            num_instances = i+1;
+        }
+    }
+}
+
+// update WheelEncoder state for all instances. This should be called by main loop
+void AP_WheelEncoder::update(void)
+{
+    for (uint8_t i=0; i<num_instances; i++) {
+        if (drivers[i] != nullptr && _type[i] != WheelEncoder_TYPE_NONE) {
+            drivers[i]->update();
+        }
+    }
+}
+    
+// check if an instance is healthy
+bool AP_WheelEncoder::healthy(uint8_t instance) const
+{
+    if (instance >= num_instances) {
+        return false;
+    }
+    return true;
+}
+
+// check if an instance is activated
+bool AP_WheelEncoder::enabled(uint8_t instance) const
+{
+    if (instance >= num_instances) {
+        return false;
+    }
+    // if no sensor type is selected, the sensor is not activated.
+    if (_type[instance] == WheelEncoder_TYPE_NONE) {
+        return false;
+    }
+    return true;
+}
+
+// get the total distance travelled in meters
+Vector2f AP_WheelEncoder::get_position(uint8_t instance) const
+{
+    // for invalid instances return zero vector
+    if (instance >= WHEELENCODER_MAX_INSTANCES) {
+        return Vector2f(0.0f, 0.0f);
+    }
+    return Vector2f(_pos_x[instance],_pos_y[instance]);
+}
+
+
+// get the total distance traveled in meters
+float AP_WheelEncoder::get_distance(uint8_t instance) const
+{
+    // for invalid instances return zero
+    if (instance >= WHEELENCODER_MAX_INSTANCES) {
+        return 0.0f;
+    }
+    return _scaling[instance] * state[instance].distance_count * 0.001f;
+}
+
+// get the total number of sensor reading from the encoder
+uint32_t AP_WheelEncoder::get_total_count(uint8_t instance) const
+{
+    // for invalid instances return zero
+    if (instance >= WHEELENCODER_MAX_INSTANCES) {
+        return 0;
+    }
+    return state[instance].total_count;
+}
+
+// get the total distance traveled in meters
+uint32_t AP_WheelEncoder::get_error_count(uint8_t instance) const
+{
+    // for invalid instances return zero
+    if (instance >= WHEELENCODER_MAX_INSTANCES) {
+        return 0;
+    }
+    return state[instance].error_count;
+}
+
+// get the signal quality for a sensor
+float AP_WheelEncoder::get_signal_quality(uint8_t instance) const
+{
+    // protect against divide by zero
+    if (state[instance].total_count == 0) {
+        return 0.0f;
+    }
+    return constrain_float((1.0f - ((float)state[instance].error_count / (float)state[instance].total_count)) * 100.0f, 0.0f, 100.0f);
```

Comment:
Yep I didn't make the calculation ... 200km is fine, your motor died before !
About the period, yes we can probably improve the time windows with some better statistic, but it is a futur enhancement